File: app/src/main/java/com/jakebarnby/filemanager/util/ComparatorUtils.java
package com.jakebarnby.filemanager.util;

import android.content.Context;

import com.jakebarnby.filemanager.sources.models.SourceFile;

import java.util.Comparator;

/**
 * Created by Jake on 10/3/2017.
 */

public class ComparatorUtils {
    public static Comparator<TreeNode<? extends SourceFile>> resolveComparator(Context context) {
        boolean showFoldersFirst = PreferenceUtils.getBoolean(
                context,
                Constants.Prefs.FOLDER_FIRST_KEY,
                true);

        int sortType = PreferenceUtils.getInt(
                context,
                Constants.Prefs.SORT_TYPE_KEY,
                Constants.SortTypes.NAME);

        int orderType = PreferenceUtils.getInt(
                context,
                Constants.Prefs.ORDER_TYPE_KEY,
                Constants.SortTypes.NAME);

        switch (sortType) {
            case Constants.SortTypes.NAME:
                return getComparatorNameForOrder(orderType, showFoldersFirst);
            case Constants.SortTypes.MODIFIED_TIME:
                return getComparatorTimeForOrder(orderType, showFoldersFirst);
            case Constants.SortTypes.SIZE:
                return getComparatorSizeForOrder(orderType, showFoldersFirst);
            case Constants.SortTypes.TYPE:
                return getComparatorTypeForOrder(orderType, showFoldersFirst);
            default:
                return getComparatorNameForOrder(orderType, showFoldersFirst);

        }
    }

    private static Comparator<TreeNode<? extends SourceFile>> getComparatorNameForOrder(int orderType, boolean showFoldersFirst) {
        if (orderType == Constants.OrderTypes.ASCENDING) {
            return (node1, node2) -> {
                if (showFoldersFirst) {
                    int result = Boolean.valueOf(!node1.getData().isDirectory()).compareTo(!node2.getData().isDirectory());
                    if (result == 0) {
                        result = node1.getData().getName().toLowerCase().compareTo(node2.getData().getName().toLowerCase());
                    }
                    return result;
                } else {
                    return node1.getData().getName().toLowerCase().compareTo(node2.getData().getName().toLowerCase());
                }
            };
        } else {
            return (node1, node2) -> {
                if (showFoldersFirst) {
                    int result = Boolean.valueOf(!node1.getData().isDirectory()).compareTo(!node2.getData().isDirectory());
                    if (result == 0) {
                        result = node2.getData().getName().toLowerCase().compareTo(node1.getData().getName().toLowerCase());
                    }
                    return result;
                } else {
                    return node2.getData().getName().toLowerCase().compareTo(node1.getData().getName().toLowerCase());
                }
            };
        }
    }


    private static Comparator<TreeNode<? extends SourceFile>> getComparatorTimeForOrder(int orderType, boolean showFoldersFirst) {
        if (orderType == Constants.OrderTypes.ASCENDING) {
            return (node1, node2) -> {
                if (showFoldersFirst) {
                    int result = Boolean.valueOf(!node1.getData().isDirectory()).compareTo(!node2.getData().isDirectory());
                    if (result == 0) {
                        result = Long.compare(node1.getData().getModifiedTime(), (node2.getData().getModifiedTime()));
                    }
                    return result;
                } else {
                    return Long.compare(node1.getData().getModifiedTime(), (node2.getData().getModifiedTime()));
                }
            };
        } else {
            return (node1, node2) -> {
                if (showFoldersFirst) {
                    int result = Boolean.valueOf(!node1.getData().isDirectory()).compareTo(!node2.getData().isDirectory());
                    if (result == 0) {
                        result = Long.compare(node2.getData().getModifiedTime(), (node1.getData().getModifiedTime()));
                    }
                    return result;
                } else {
                    return Long.compare(node2.getData().getModifiedTime(), (node1.getData().getModifiedTime()));
                }
            };
        }
    }

    private static Comparator<TreeNode<? extends SourceFile>> getComparatorSizeForOrder(int orderType, boolean showFoldersFirst) {
        if (orderType == Constants.OrderTypes.ASCENDING) {
            return (node1, node2) -> {
                if (showFoldersFirst) {
                    int result = Boolean.valueOf(!node1.getData().isDirectory()).compareTo(!node2.getData().isDirectory());
                    if (result == 0) {
                        result = Long.compare(node1.getData().getSize(), (node2.getData().getSize()));
                    }
                    return result;
                } else {
                    return Long.compare(node1.getData().getSize(), (node2.getData().getSize()));
                }
            };
        } else {
            return (node1, node2) -> {
                if (showFoldersFirst) {
                    int result = Boolean.valueOf(!node1.getData().isDirectory()).compareTo(!node2.getData().isDirectory());
                    if (result == 0) {
                        result = Long.compare(node2.getData().getSize(), (node1.getData().getSize()));
                    }
                    return result;
                } else {
                    return Long.compare(node2.getData().getSize(), (node1.getData().getSize()));
                }
            };
        }
    }

    private static Comparator<TreeNode<? extends SourceFile>> getComparatorTypeForOrder(int orderType, boolean showFoldersFirst) {
        if (orderType == Constants.OrderTypes.ASCENDING) {
            return (node1, node2) -> {
                if (showFoldersFirst) {
                    int result = Boolean.valueOf(!node1.getData().isDirectory()).compareTo(!node2.getData().isDirectory());
                    if (result == 0) {
                        result = Long.compare(node1.getData().getSize(), (node2.getData().getSize()));
                    }
                    return result;
                } else {
                    return Long.compare(node1.getData().getSize(), (node2.getData().getSize()));
                }
            };
        } else {
            return (node1, node2) -> {
                if (showFoldersFirst) {
                    int result = Boolean.valueOf(!node1.getData().isDirectory()).compareTo(!node2.getData().isDirectory());
                    if (result == 0) {
                        result = Long.compare(node2.getData().getSize(), (node1.getData().getSize()));
                    }
                    return result;
                } else {
                    return Long.compare(node2.getData().getSize(), (node1.getData().getSize()));
                }
            };
        }
    }
}

File: app/src/main/java/com/jakebarnby/filemanager/util/ComparatorUtils.kt
package com.jakebarnby.filemanager.util

import android.content.Context
import com.jakebarnby.filemanager.sources.models.SourceFile
import com.jakebarnby.filemanager.util.Constants.OrderTypes
import com.jakebarnby.filemanager.util.Constants.Prefs
import com.jakebarnby.filemanager.util.Constants.SortTypes
import java.util.*

/**
 * Created by Jake on 10/3/2017.
 */
object ComparatorUtils {

    fun resolveComparatorForPrefs(context: Context): Comparator<TreeNode<out SourceFile>> {
        val showFoldersFirst = PreferenceUtils.getBoolean(
                context,
                Prefs.FOLDER_FIRST_KEY,
                true
        )
        val sortType = PreferenceUtils.getInt(
                context,
                Prefs.SORT_TYPE_KEY,
                SortTypes.NAME
        )
        val orderType = PreferenceUtils.getInt(
                context,
                Prefs.ORDER_TYPE_KEY,
                SortTypes.NAME
        )

        return when (sortType) {
            SortTypes.NAME ->
                getComparatorNameForOrder(orderType, showFoldersFirst)
            SortTypes.MODIFIED_TIME ->
                getComparatorTimeForOrder(orderType, showFoldersFirst)
            SortTypes.SIZE ->
                getComparatorSizeForOrder(orderType, showFoldersFirst)
            SortTypes.TYPE ->
                getComparatorTypeForOrder(orderType, showFoldersFirst)
            else ->
                getComparatorNameForOrder(orderType, showFoldersFirst)
        }
    }

    private fun getComparatorNameForOrder(orderType: Int, showFoldersFirst: Boolean): Comparator<TreeNode<out SourceFile>> {
        return if (orderType == OrderTypes.ASCENDING) {
            Comparator { node1, node2 ->
                if (showFoldersFirst) {
                    var result = (!node1.data.isDirectory).compareTo(!node2.data.isDirectory)
                    if (result == 0) {
                        result = node1.data.name.compareTo(node2.data.name, true)
                    }
                    return@Comparator result
                } else {
                    return@Comparator node1.data.name.compareTo(node2.data.name, true)
                }
            }
        } else {
            Comparator { node1: TreeNode<out SourceFile>, node2: TreeNode<out SourceFile> ->
                if (showFoldersFirst) {
                    var result = (!node1.data.isDirectory).compareTo(!node2.data.isDirectory)
                    if (result == 0) {
                        result = node2.data.name.compareTo(node1.data.name, true)
                    }
                    return@Comparator result
                } else {
                    return@Comparator node2.data.name.compareTo(node1.data.name, true)
                }
            }
        }
    }

    private fun getComparatorTimeForOrder(orderType: Int, showFoldersFirst: Boolean): Comparator<TreeNode<out SourceFile>> {
        return if (orderType == OrderTypes.ASCENDING) {
            Comparator { node1: TreeNode<out SourceFile>, node2: TreeNode<out SourceFile> ->
                if (showFoldersFirst) {
                    var result = (!node1.data.isDirectory).compareTo(!node2.data.isDirectory)
                    if (result == 0) {
                        result = node1.data.modifiedTime.compareTo(node2.data.modifiedTime)
                    }
                    return@Comparator result
                } else {
                    return@Comparator node1.data.modifiedTime.compareTo(node2.data.modifiedTime)
                }
            }
        } else {
            Comparator { node1: TreeNode<out SourceFile>, node2: TreeNode<out SourceFile> ->
                if (showFoldersFirst) {
                    var result = (!node1.data.isDirectory).compareTo(!node2.data.isDirectory)
                    if (result == 0) {
                        result = node2.data.modifiedTime.compareTo(node1.data.modifiedTime)
                    }
                    return@Comparator result
                } else {
                    return@Comparator node2.data.modifiedTime.compareTo(node1.data.modifiedTime)
                }
            }
        }
    }

    private fun getComparatorSizeForOrder(orderType: Int, showFoldersFirst: Boolean): Comparator<TreeNode<out SourceFile>> {
        return if (orderType == OrderTypes.ASCENDING) {
            Comparator { node1: TreeNode<out SourceFile>, node2: TreeNode<out SourceFile> ->
                if (showFoldersFirst) {
                    var result = (!node1.data.isDirectory).compareTo(!node2.data.isDirectory)
                    if (result == 0) {
                        result = node1.data.size.compareTo(node2.data.size)
                    }
                    return@Comparator result
                } else {
                    return@Comparator node1.data.size.compareTo(node2.data.size)
                }
            }
        } else {
            Comparator { node1: TreeNode<out SourceFile>, node2: TreeNode<out SourceFile> ->
                if (showFoldersFirst) {
                    var result = (!node1.data.isDirectory).compareTo(!node2.data.isDirectory)
                    if (result == 0) {
                        result = node2.data.size.compareTo(node1.data.size)
                    }
                    return@Comparator result
                } else {
                    return@Comparator node2.data.size.compareTo(node1.data.size)
                }
            }
        }
    }

    private fun getComparatorTypeForOrder(orderType: Int, showFoldersFirst: Boolean): Comparator<TreeNode<out SourceFile>> {
        return if (orderType == OrderTypes.ASCENDING) {
            Comparator { node1: TreeNode<out SourceFile>, node2: TreeNode<out SourceFile> ->
                if (showFoldersFirst) {
                    var result = (!node1.data.isDirectory).compareTo(!node2.data.isDirectory)
                    if (result == 0) {
                        result = node1.data.size.compareTo(node2.data.size)
                    }
                    return@Comparator result
                } else {
                    return@Comparator node1.data.size.compareTo(node2.data.size)
                }
            }
        } else {
            Comparator { node1: TreeNode<out SourceFile>, node2: TreeNode<out SourceFile> ->
                if (showFoldersFirst) {
                    var result = (!node1.data.isDirectory).compareTo(!node2.data.isDirectory)
                    if (result == 0) {
                        result = node2.data.size.compareTo(node1.data.size)
                    }
                    return@Comparator result
                } else {
                    return@Comparator node2.data.size.compareTo(node1.data.size)
                }
            }
        }
    }
}